### PR TITLE
onFocus and onBlur support

### PR DESCRIPTION
**Pull request by YeungKC**
## Description

Make EditableText, CupertinoTextField and TextField support onFocus and onBlur.

## Related Issues

#9728

## Tests

- editable_text_test.dart
  - I added the following tests:

Call onFocus and onBlur correctly 

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

File: packages/flutter/test/widgets/editable_text_test.dart [hunk 1]
```diff
@@ -6476,6 +6476,81 @@ void main() {
...
+              onBlur: () {
+                onBlurBarCalledTime += 1;
+              },
+            ),
+          ],
+        ),
+      ),
+    );
+
+
```
Comment (justinmc) [inline packages/flutter/test/widgets/editable_text_test.dart:6526]: Too many newlines here.

File: packages/flutter/lib/src/cupertino/text_field.dart [hunk 2]
```diff
@@ -258,6 +258,8 @@ class CupertinoTextField extends StatefulWidget {
     this.onChanged,
     this.onEditingComplete,
     this.onSubmitted,
+    this.onFocus,
+    this.onBlur,
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/text_field.dart:262]: I recall a team discussion where it was decided that "unfocus" is superior to "blur".  Should this be "onUnfocus" @HansMuller?

**Review by justinmc — COMMENTED** [review]
The PR's approach looks good, but I want to make sure that we get buy-in on adding these new parameters.

The concept of onFocus and onBlur can currently already be handled by using a FocusNode, so these new methods are just convenience.  However, as mentioned in the [issue](https://github.com/flutter/flutter/issues/9728), this is a common pattern that users will likely expect to exist, and it's quite verbose to implement their behavior right now.

Currently to get this behavior, you'd have to maintain a listener to the FocusNode in a StatefulWidget and maintain a boolean with the previous value of `hasFocus`.  I think I agree that this is verbose and error-prone enough to merit these two new callbacks, but what does everyone else think?

CC'ing some people that might have thoughts: @HansMuller @LongCatIsLooong @xu-baolin @jamesblasco

**Comment by LongCatIsLooong** [comment]
Also there is a `Focus` widget that provides an `onFocusChanged` callback:
```dart
Focus(
  onFocusChanged: (value) {
    ...
  },
  child: Builder(BuildContext context) => TextField(focusNode: Focus.of(context)),
);
```

**Comment by YeungKC** [comment]
It looks pretty good too, but I'd feel better if `focusNode: Focus.of(context)` was the default behavior.

**Comment by YeungKC** [comment]
I didn't know about Focus before, so I think it's ok to close this pr if Focus exists.

But I still think it's better if the default value of the TextField focusNode is Focus.of(context).

Maybe I should open a new issue to discuss this.

cc @justinmc  @LongCatIsLooong

**Comment by LongCatIsLooong** [comment]
I think `TextField` needs its own `FocusNode` (being a leaf node and interactive). If you look up a `Focus` widget in the `BuildContext` using `Focus.of` you may find a `Focus` that's meant for some other widgets I think? For instance you can have a `Focus` widget for your form (probably a bad example). If the descendant text fields use `Focus.of` to retrieve a focus node then they will end up sharing the same focus node that's meant for the form.

**Comment by YeungKC** [comment]
After thinking about it, I do think it's a little different than other InheritedWidgets, so thanks for the answer.

**Pull request closed without merging**